File: bench/cross-language/matrix_mult/matrix_mult.zig
const std = @import("std");

const N = 200;

pub fn main() !void {
    const stdout_file = std.fs.File.stdout();
    var buf: [4096]u8 = undefined;
    var w = stdout_file.writer(&buf);

    // Use arrays on the heap via page allocator since 200x200 is large for stack
    var a: [N][N]f64 = undefined;
    var b: [N][N]f64 = undefined;
    var c: [N][N]f64 = undefined;

    // Initialize matrices
    for (0..N) |i| {
        for (0..N) |j| {
            a[i][j] = @as(f64, @floatFromInt((i * N + j) % 1000)) / 1000.0;
            b[i][j] = @as(f64, @floatFromInt((j * N + i) % 1000)) / 1000.0;
            c[i][j] = 0.0;
        }
    }

    // Multiply C = A * B
    for (0..N) |i| {
        for (0..N) |j| {
            var sum: f64 = 0.0;
            for (0..N) |k| {
                sum += a[i][k] * b[k][j];
            }
            c[i][j] = sum;
        }
    }

    // Checksum
    var checksum: f64 = 0.0;
    for (0..N) |i| {
        for (0..N) |j| {
            checksum += c[i][j];
        }
    }

    try w.interface.print("matrix_mult(200): checksum = {d:.6}\n", .{checksum});
    try w.interface.flush();
}

File: bench/cross-language/matrix_mult/matrix_mult.go
package main

import "fmt"

const N = 200

func main() {
	var A [N][N]float64
	var B [N][N]float64
	var C [N][N]float64

	// Initialize matrices
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			A[i][j] = float64((i*N+j)%1000) / 1000.0
			B[i][j] = float64((j*N+i)%1000) / 1000.0
			C[i][j] = 0.0
		}
	}

	// Multiply C = A * B
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			sum := 0.0
			for k := 0; k < N; k++ {
				sum += A[i][k] * B[k][j]
			}
			C[i][j] = sum
		}
	}

	// Checksum
	checksum := 0.0
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			checksum += C[i][j]
		}
	}

	fmt.Printf("matrix_mult(200): checksum = %.6f\n", checksum)
}
